crates/bifrost/src/providers/replicated_loglet/rpc_routers.rs

```diff
@@ -55,15 +73,600 @@ impl LogServersRpc {
+                            // we retry to re-connect one time
+                            stream = self.stream.renew(loglet.networking(), &stream).await?;
+                        }
+                        err => return Err(err),
+                    }
+                }
+            }
+        };
+
+        Ok(commit_token)
+    }
+}
+
+/// RemoteSequencerStream represents a single open connection
+/// to remote leader sequencer.
+///
+/// This connection handles all [`Appended`] responses from the remote
+/// sequencer.
+///
+/// If the connection was lost or if any of the commits failed
+/// with a terminal error (like [`Status::Disabled`]) all pending commits
+/// are resolved with an error.
+#[derive(Clone)]
+struct RemoteSequencerConnection {
+    rpc_router: RpcRouter<Append>,
+    connection: WeakConnection,
+    tx: mpsc::UnboundedSender<RemoteSequencerWaitTask>,
+}
+
+impl RemoteSequencerConnection {
+    fn start(
+        rpc_router: RpcRouter<Append>,
+        connection: WeakConnection,
+    ) -> Result<Self, ShutdownError> {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        task_center().spawn(
+            TaskKind::Disposable,
+            "remote-sequencer-connection",
+            None,
+            Self::handle_appended_responses(rx),
+        )?;
+
+        Ok(Self {
+            rpc_router,
+            connection,
+            tx,
+        })
+    }
+
+    /// Send append message to remote sequencer.
+    ///
+    /// It's up to the caller to retry on [`NetworkError`]
+    pub async fn send(
+        &self,
+        sequencer: GenerationalNodeId,
+        permit: OwnedSemaphorePermit,
+        msg: Append,
+    ) -> Result<LogletCommit, NetworkError> {
+        if self.tx.is_closed() {
+            return Err(NetworkError::ConnectionClosed);
+        }
+
+        let outgoing = Outgoing::new(sequencer, msg).assign_connection(self.connection.clone());
+
+        let rpc_token = self
+            .rpc_router
+            .send_on_connection(outgoing)
+            .await
+            .map_err(|err| err.source)?;
+
+        let (commit, commit_resolver) = LogletCommit::deferred();
+        let task = RemoteSequencerWaitTask {
+            rpc_token,
+            commit_resolver,
+            permit,
+        };
+
+        // wait for response asynchronously
+        self.tx
+            .send(task)
+            .map_err(|_| NetworkError::ConnectionClosed)?;
+
+        Ok(commit)
+    }
+
+    /// Handle all [`Appended`] responses
+    ///
+    /// This task will run until the [`AppendStream`] is dropped. Once dropped
+    /// all pending commits will be resolved with an error. it's up to the enqueuer
+    /// to retry if needed.
+    async fn handle_appended_responses(
+        mut rx: mpsc::UnboundedReceiver<RemoteSequencerWaitTask>,
+    ) -> anyhow::Result<()> {
+        let mut waiting = FuturesUnordered::new();
+        let cancel = CancellationToken::new();
```

Comment:
I am considering using a `Broadcast` channel instead of a cancellation token because I need to communicate the `reason` for cancellation as well. Not the fact that it was only cancelled.

Then we can cancel all pending tokens with proper error/sealed, etc. instead of just connection closed as I do now.